pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +170,122 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
+	nodeIPtoBlock, pv, err := c.getWatcherIpPortToBlock(ctx, node, clusterName)
+	if err != nil {
+		return fmt.Errorf("failed to get node IP to block for node %q. %v", node.Name, err)
+	}
+	fmt.Println("NodeIP to block", nodeIPtoBlock)
+
+	networkFence := &addonsv1alpha1.NetworkFence{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      node.Name,
+			Namespace: c.namespace,
+		},
+		Spec: addonsv1alpha1.NetworkFenceSpec{
+			Driver:     "rook-ceph.rbd.csi.ceph.com",
+			FenceState: addonsv1alpha1.Fenced,
+			Secret: addonsv1alpha1.SecretSpec{
+				Name:      pv.Spec.CSI.ControllerPublishSecretRef.Name,
+				Namespace: pv.Spec.CSI.ControllerPublishSecretRef.Namespace,
+			},
+			Cidrs: []string{
+				nodeIPtoBlock,
+			},
+			Parameters: map[string]string{
+				"clusterID": pv.Spec.CSI.VolumeAttributes["clusterID"],
+			},
+		},
+	}
+
+	err = c.client.Create(ctx, networkFence)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			logger.Infof("network fence CR %s already exists", node.Name)
+		} else {
+			logger.Errorf("failed to create network fence CR. %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *clientCluster) deleteNetworkFence(ctx context.Context, node corev1.Node) error {
+	err := c.client.Delete(ctx, &addonsv1alpha1.NetworkFence{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      node.Name,
+			Namespace: c.namespace,
+		},
+	})
+	if err != nil && !errors.IsNotFound(err) {
+		logger.Errorf("failed to delete network fence CR. %v", err)
+		return err
+	}
+	return nil
+}
+
+func (c *clientCluster) getWatcherIpPortToBlock(ctx context.Context, node corev1.Node, clusterName string) (string, corev1.PersistentVolume, error) {
+	volumesInuse := node.Status.VolumesInUse
+	logger.Debugf("volumesInuse %s", volumesInuse)
+
+	listPVs, err := c.context.Clientset.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", corev1.PersistentVolume{}, fmt.Errorf("failed to list Cephfs PV. %v", err)
+	}
+
+	var listRbdPV, listCephFSPV []corev1.PersistentVolume
+
+	for _, pv := range listPVs.Items {
+		if pv.Spec.CSI.Driver == "rook-ceph.rbd.csi.ceph.com" {
+			listRbdPV = append(listRbdPV, pv)
+		} else if pv.Spec.CSI.Driver == "rook-ceph.cephfs.csi.ceph.com" {
+			listCephFSPV = append(listCephFSPV, pv)
+		}
+	}
+
+	if len(volumesInuse) == 0 || (len(listRbdPV) == 0 && len(listCephFSPV) == 0) {
+		logger.Error("No PVs found on the node")
+		return "", corev1.PersistentVolume{}, nil
+	}
+
+	clusterInfo := cephclient.AdminClusterInfo(ctx, c.namespace, clusterName)
+
+	for _, volume := range volumesInuse {
+		volumesInuseRemoveK8sPrefix := strings.TrimPrefix(string(volume), "kubernetes.io/csi/")
+		splitVolumeInUseBased := strings.Split(volumesInuseRemoveK8sPrefix, "^")
+		logger.Infof("splitVolumeInUseBased %v", splitVolumeInUseBased)
+		if len(splitVolumeInUseBased) == 2 && splitVolumeInUseBased[0] == "rook-ceph.rbd.csi.ceph.com" {
```

Comment:
* get the corresponding pv`s for the volumes in use, we should not run rbd commands for all the PVC.
* Try to get the IP for the rbd image. Once you find IP, could you return from here?

Reply:
question on second part
```
is return after first list of watchers is okay since there are multiple volume may be attached to a node and ip os watcher could be different or not?
```